<a href="https://colab.research.google.com/github/Omar-RojasGBF/lis5693/blob/main/lab-2/Lab_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install spacy
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 92.2 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [2]:
import spacy

I installed spaCy and the small English model to get started on some basic text pre-processing.

In [10]:
import requests

url = "https://raw.githubusercontent.com/Omar-RojasGBF/lis5693/refs/heads/main/lab-1/transcript.txt"
response = requests.get(url)
response.raise_for_status()
text = response.text

In [13]:
print("Number of characters:", len(text))

print(text[100])

Number of characters: 8566
 


I loaded and read my transcript.txt file from my lab-1 GitHub repo. There are 8566 characters, and I loaded the first 100 characters.

In [14]:
nlp = spacy.blank("en")
nlp.add_pipe("sentencizer")

doc = nlp(text)

sentences = list(doc.sents)

print("Number of sentences:", len(sentences))

print("\nFirst 3 sentences:")
for sent in sentences[:3]:
  print(sent)

Number of sentences: 1

First 3 sentences:
hello I'm blue and in today's video 
I'll be going over color theory  
but before I start this video is brought to you by 
the kind people over at skillshare skillshare is  
an online learning platform where you can learn 
things from photography to Art how to manage a  
career in the creative fields and finances and 
so much more skillshare is a great class that  
deep dive into the intricacy of color theory 
and color picking color palettes which are  
told by professional artists from classes like 
one poet illustrator discover how Color Works  
in your art by Tom Forest to the color Survival 
Guide by Marco Busey but these classes of unique  
perspective to use of colors there's so many 
more classes touching another subject too with  
unique perspectives and insight not just in the 
creative and physical side of learning skillshare  
also offers classes on developing and choosing 
career paths from working in the creative field  
to manag

I added a sentencizer to a blank spaCy mdoel. There is 1 sentence in my transcript. The transcript seemed to ignore all periods and other forms of grammar.

In [15]:
words = [token.text.lower() for token in doc if token.is_alpha]

print("Total words:", len(words))
print("Unique words:", len(set(words)))

Total words: 1543
Unique words: 430


There are 1543 words in total, and 430 unique words.

In [16]:
from collections import Counter

word_freq = Counter(words)

print("Top 10 most frequent words:")
for word, count in word_freq.most_common(10):
  print(word, count)

Top 10 most frequent words:
the 87
color 57
and 56
to 55
is 44
a 36
of 35
colors 35
i 30
in 26


The most frequent non-stopword is "color." This video is about color theory, so I would assume that the word "color" would come up alot, since the speaker could refer to terms such as the color wheel, color swatches, or color palettes.

In [17]:
nlp2 = spacy.load("en_core_web_sm")

doc2 = nlp2(text)

print("Named Entities:")
for ent in doc2.ents:
  print(ent.text, ent.label_)

Named Entities:
today DATE
Art PERSON
one CARDINAL
Color Works PERSON
Tom Forest PERSON
Survival GPE
Marco Busey PERSON
one month DATE
first ORDINAL
thousand CARDINAL
today DATE
two CARDINAL
the Outer Circle WORK_OF_ART
Focus ORG
100 CARDINAL
two CARDINAL
50 50 CARDINAL
two CARDINAL
Deep NORP
two CARDINAL
one CARDINAL
two CARDINAL
Hues ORG
one CARDINAL
Hues ORG
two CARDINAL
two CARDINAL
Hues PERSON
Vibes ORG
Hues PERSON
Hues PERSON
Hues PERSON
Keys PERSON


I ran a full spaCy pipeline. There were 9 named entities found. The types of entities that appeared were PERSON, ORG, CARDINAL, ORG, NORP, WORK_OF_ART, DATE, ORDINAL, and GPE.

In [18]:
from spacy.matcher import PhraseMatcher

matcher = PhraseMatcher(nlp2.vocab, attr="LOWER")

phrases = ["color scheme", "color theory", "hues"]

patterns = [nlp2(p) for p in phrases]

matcher.add("TECH_TERMS", patterns)

matches = matcher(doc2)

print("Matches found:")
for match_id, start, end in matches:
  print(doc2[start:end])
  print("Sentence:", doc2[start].sent)

Matches found:
color theory
Sentence: hello I'm blue and in today's video 
I'll be going over color theory  
but before I start this video is brought to you by 
the kind people over at skillshare skillshare is  
an online learning platform where you can learn 
things from photography to Art how to manage a  
career in the creative fields and finances and 
so much more skillshare is a great class that  
deep dive into the intricacy of color theory 
and color picking color palettes which are  
told by professional artists from classes like 
one poet illustrator discover how Color Works  
in your art by Tom Forest to the color Survival 
Guide by Marco Busey but these classes of unique  
perspective to use of colors there's so many 
more classes touching another subject too with  
unique perspectives and insight not just in the 
creative and physical side of learning skillshare  
also offers classes on developing and choosing 
career paths from working in the creative field  
to managing f

Reflection 💭


---
I would say that everything went well. The only thing that was a bit off was the sentence segmentation when it returned only 1 sentence. I did not realize that my transcript.txt file did not take grammar into account. I guess this is a possible consequence of letting an AI generate its own transcript with YouTube's automatic speech recognition technology. It is also not showing me the 100 first characters that I wanted to print when I loaded and read my GitHub file. In general, spaCy was a great way for me to learn more about some basic text pre-processing, I was just not expecting for my transcript to only count as one sentence. I wonder if other videos with auto-transcripts have the same issue.
